Sources/PromiseKit/Future.swift

```diff
@@ -24,10 +24,11 @@ public extension Future {
   
   func setResult(_ result: FutureResult?, on dispatchQueue: DispatchQueue? = nil) {
     write {
+      guard self.internalResult == nil else { return }
       self.internalResult = result
       guard self.isEnabled, let result = result else { return }
-      dispatchQueue.async {
-        self.observers.forEach { $0.notifity(result) }
+      for observer in self.observers {
+        dispatchQueue.async { observer.notifity(result) }
```

Comment:
The root of the problem is here. By keeping the loop on the thread that resolved the promise and only dispatching the observers themselves, tests work as expected.